### PR TITLE
Fix bugs when exporting MD Histo cuts workspaces in sliceviewer

### PR DESCRIPTION
**Pull request by MohamedAlmaki**
### Description of work
This PR fixes several bugs that happen when exporting 3D MD Histo workspaces in sliceviewer. For more details check #36084

Fixes part of #36084. 

#### Further detail of work

The issue was that `bin_params` is None when the workspace is MD Histo workspace. I handled this case by setting the bin width to the dim bin width when the `bin_params` is None. 
### To test:
Follow #36084

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

### CODE REVIEW

**Comment by sf1919** [comment]
This is still in draft so I assume that there are more commits to come.

Can you keep an eye on the Doxygen build? It is not failing if there are warnings and currently there are 7 on this build. Those for [LoadBankFromDiskTask.cpp](https://builds.mantidproject.org/job/pull_requests-doxygen/52093/doxygen/fileName.716520925/) and [WorkspaceUtils.cpp](https://builds.mantidproject.org/job/pull_requests-doxygen/52093/doxygen/fileName.1095366228/) are being dealt with separately but I don't recognise [-](https://builds.mantidproject.org/job/pull_requests-doxygen/52093/doxygen/fileName.45/) as one of the ones being resolved. It may therefore be specific to this PR.

**Review by RichardWaiteSTFC — CHANGES_REQUESTED** [review]
Thanks for this, fix looks good and it resolves issue (1) in #36084 (exporting ROI) but not issue (2) (exporting cuts).
The error message says
```
SliceViewerModel.export_cuts_to_workspace_mdevent() missing 1 required positional argument: 'cut'
```
But it should in fact be calling `self.export_cuts_to_workspace_mdhisto` - if you have time to look into it that'd be great, if not then perhaps could you add a unit test for this and then I'll approve!

**Comment by MohamedAlmaki** [comment]
I couldn't reproduce 2 after multiple tries. The exporting is working fine using all types of cuts. Could you please send me the file you used to produce 2 or a different script than the one in the issue?

**Comment by RichardWaiteSTFC** [comment]
Just had another look at this - so exporting cuts does work if you use the rectangle ROI tool, but if you just open line plots (i.e. cuts through the data at the cursor position) i get the error I posted above
`SliceViewerModel.export_cuts_to_workspace_mdevent() missing 1 required positional argument: 'cut'`
I'll keep digging into it, but at least hopefully you can reproduce this now!

**Comment by RichardWaiteSTFC** [comment]
So the bug is here in `export_pixel_cut_to_workspace_md`
https://github.com/mantidproject/mantid/blob/42f0fb5940f8197a354d834668502bf740644d4d/qt/python/mantidqt/mantidqt/widgets/sliceviewer/models/model.py#L477
If this returns `self.export_cut_to_workspace` then we at least get one step closer - now there is a missing argument`dimensions_indices` (despite what the error says).
If you add in `dimension_indices=data_view.dimensions.get_states()` in the call to `self.model.export_pixel_cut_to_workspace` here
https://github.com/mantidproject/mantid/blob/42f0fb5940f8197a354d834668502bf740644d4d/qt/python/mantidqt/mantidqt/widgets/sliceviewer/presenters/presenter.py#L292
and alter the function definitions in the model accordingly you no ;longer get an error but the behaviour isn't quite right. 
For an MDEvent workspace it seems to get the integration limits as double the bin width (i.e. twice what it should be), for an MDHisto workspace the cut is only 2 bins wide (i.e. 3  bins included along the axis of the cut as opposed to the limits of the data).
Hope that helps!

**Review by RichardWaiteSTFC — CHANGES_REQUESTED** [review]
As discussed offline, we should merge this fix for issue (1) and do issue (2) in a separate PR as it's proving difficult. Can you add a release note and make a note in the original issue?

Also is it possible to look into why this tests
https://github.com/mantidproject/mantid/blob/3d69faf7636eed5a4d1fa6ef224c79caa8f23c73/qt/python/mantidqt/mantidqt/widgets/sliceviewer/test/test_sliceviewer_model.py#L682
Isn't failing - it might be a problem with the test or maybe we need another test to stop this regression?
Thanks!

File: qt/python/mantidqt/mantidqt/widgets/sliceviewer/models/model.py [hunk 1]
```diff
@@ -676,12 +676,14 @@ def _dimension_limits(
     :param limits: An optional Sequence of length 2 containing limits for plotting dimensions. If
                     not provided the full extent of each dimension is used.
     """
-    dim_limits = [(dim.getMinimum(), dim.getMaximum()) for dim in [workspace.getDimension(i) for i in range(workspace.getNumDims())]]
+    dims = [workspace.getDimension(i) for i in range(workspace.getNumDims())]
+    dim_limits = [(dim.getMinimum(), dim.getMaximum()) for dim in dims]
+    dim_width = [dim.getBinWidth() for dim in dims]
```
Comment (RichardWaiteSTFC) [inline qt/python/mantidqt/mantidqt/widgets/sliceviewer/models/model.py:681]: I suppose this comprehension could probably be done in the main loop - i.e. loop over `idim in rnage(workspace.getNumDims())` -  given we are looping over the dimensions anyway, but as there's only usually 2-4 dimensions this is pretty negligible!

Comment (MohamedAlmaki) [reply qt/python/mantidqt/mantidqt/widgets/sliceviewer/models/model.py:681]: Hmm, why do you think the test should fail? the test doesn't cover the current new addition as `bin_params` is not None and the latest addition handles the case when `bin_params` is None

Ok, I will refactor the code, thanks for spotting the duplication

Comment (RichardWaiteSTFC) [reply qt/python/mantidqt/mantidqt/widgets/sliceviewer/models/model.py:681]: Thanks, I see why the test didn't fail, but I think that indicates we need another test for when `bin_params` is `None`

**Review by RichardWaiteSTFC — APPROVED** [review]
Looks good thanks for fixing this!

**Pull request merged**